triage/experiments/validate.py

```diff
@@ -0,0 +1,221 @@
+            raise ValueError("configuration key 'feature_aggregations' not found")
+        for aggregation in feature_aggregation_config:
+            self._validate_aggregation(aggregation)
+
+class EventsTableValidator(Validator):
+    def run(self, events_table):
+        if not events_table:
+            raise ValueError("configuration key 'events_table' not found")
+        table_should_have_data(events_table, self.db_engine)
+        column_should_be_intlike(events_table, 'entity_id', self.db_engine)
+        column_should_be_timelike(events_table, 'outcome_date', self.db_engine)
+        column_should_be_booleanlike(events_table, 'outcome', self.db_engine)
+
+
+class StateConfigValidator(Validator):
+    def run(self, state_config):
+        if 'table_name' in state_config:
+            dense_state_table = state_config['table_name']
+            table_should_have_data(dense_state_table, self.db_engine)
+            column_should_be_intlike(dense_state_table, 'entity_id', self.db_engine)
+            column_should_be_stringlike(dense_state_table, 'state', self.db_engine)
+            column_should_be_timelike(dense_state_table, 'start_time', self.db_engine)
+            column_should_be_timelike(dense_state_table, 'end_time', self.db_engine)
+            if 'state_filters' not in state_config or len(state_config['state_filters']) < 1:
+                raise ValueError('If a table_name is given in state_config, at least one state filter must be present')
+
+
+class FeatureGroupDefinitionValidator(Validator):
+    def run(self, feature_group_definition, feature_aggregation_config):
+        if not isinstance(feature_group_definition, dict):
+            raise ValueError('Feature Group Definition must be a dictionary')
+
+        available_subsetters = architect.feature_group_creator.FeatureGroupCreator.subsetters
+        for subsetter_name, value in feature_group_definition.items():
+            if subsetter_name not in available_subsetters:
+                raise ValueError('''Unknown feature_group_definition key {} received.
+                Available keys are {}'''.format(subsetter_name, available_subsetters))
+            if not hasattr(value, '__iter__') or isinstance(value, (str, bytes)):
+                raise ValueError('Each value in FeatureGroupCreator must be iterable and not a string')
+
+        if 'prefix' in feature_group_definition:
+            available_prefixes = [
+                aggregation['prefix']
+                for aggregation in feature_aggregation_config
+            ]
+            for prefix in feature_group_definition['prefix']:
+                if prefix not in available_prefixes:
+                    raise ValueError('''Aggregation prefix of '{}' was given as a
+                    feature group, but no such prefix exists in the available
+                    feature aggregations. The available feature aggregations are {}
+                    '''.format(prefix, available_prefixes))
```

Comment:
another trivial suggestion, but if you're doing a bunch of `[not] in` checks, a `set` would usually be more performant.

and, I wonder if it might be nice to tell them up front _all_ of the bad prefixes, along the lines of –

    available_prefixes = {aggregation['prefix'] for aggregation in feature_aggregation_config}
    bad_prefixes = set(feature_group_definition['prefix']) - available_prefixes
    if bad_prefixes:
        raise ValueError(….format(bad_prefixes, available_prefixes))